### PR TITLE
feat: [CDS-79267]: Add secret extractor from connector ref

### PR DESCRIPTION
**Pull request by abosii-harness**
## Describe your changes
Example usage:
```
@SimpleVisitorHelper(helperClass = SecretConnectorRefExtractorHelper.class)
public class GitStore implements GitStoreConfig, Visitable, WithConnectorRef {
...
}

for (ManifestAttributes manifestAttribute : manifestAttributes) {
      Set<VisitedSecretReference> secretReferences = manifestAttribute == null
              ? Set.of()
              : entityReferenceExtractorUtils.extractReferredSecrets(ambiance, manifestAttribute);

      if (EmptyPredicate.isNotEmpty(secretReferences)) {
        secretReferences.forEach(secretReference -> {
          secretRuntimeUsageService.createSecretRuntimeUsage(secretReference.getSecretRef(),
                  secretReference.getReferredBy(), EntityUsageDetailProto.newBuilder().build());
        });
      }
    }
```

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53722)

### CODE REVIEW

**Comment by abosii-harness** [comment]
trigger uts

**Review by ivan-83 — COMMENTED** [review]
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53722)* status: 7 file(s) to review. 0 of 2 approvals obtained. 

*[a discussion](https://reviewable.io/reviews/harness/harness-core/53722#-Ng-2uRA5Qm1OmA4aAf_:-Ng-2uRA5Qm1OmA4aAfa:b-r7f586) (no related file):*
hey @abosii-harness  this is a nice solution!
We can replace @SimpleVisitorHelper(helperClass = ConnectorRefExtractorHelper.class) with @SimpleVisitorHelper(helperClass = SecretConnectorRefExtractorHelper.class on objects that implements WithConnectorRef and call entityReferenceExtractorUtils.extractReferredSecrets(ambiance, WithConnectorRef);

Just to double confirm, I guess you have tested it, we are using @SimpleVisitorHelper annotation on the framework level for getting Set<EntityDetailProtoDTO>  for Referen By logic. 
I think we will not break anything because we are extending ConnectorRefExtractorHelper which implements the addReference method that will be called by the framework, right? 

___

**Review by abosii-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53722)* status: 7 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

*[a discussion](https://reviewable.io/reviews/harness/harness-core/53722#-Ng-2uRA5Qm1OmA4aAf_:-Ng-3pV81k5cubqIRWOz:b-b4ibmo) (no related file):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

hey @abosii-harness  this is a nice solution!
We can replace @SimpleVisitorHelper(helperClass = ConnectorRefExtractorHelper.class) with @SimpleVisitorHelper(helperClass = SecretConnectorRefExtractorHelper.class on objects that implements WithConnectorRef and call entityReferenceExtractorUtils.extractReferredSecrets(ambiance, WithConnectorRef);

Just to double confirm, I guess you have tested it, we are using @SimpleVisitorHelper annotation on the framework level for getting Set<EntityDetailProtoDTO>  for Referen By logic. 
I think we will not break anything because we are extending ConnectorRefExtractorHelper which implements the addReference method that will be called by the framework, right?
</blockquote></details>

yes, that is the idea. I checked the flow in logic for rbac and it perserve current functionality 

___

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+public class SecretConnectorRefExtractorHelper extends ConnectorRefExtractorHelper implements SecretReferenceExtractor {
+  @Named(DEFAULT_CONNECTOR_SERVICE) @Inject private ConnectorService connectorService;
+  @Inject private SecretRefInputValidationHelper secretRefInputValidationHelper;
+
+  @Override
+  public Set<VisitedSecretReference> addSecretReference(Object object, String accountIdentifier, String orgIdentifier,
+      String projectIdentifier, Map<String, Object> contextMap) {
+    final Set<EntityDetailProtoDTO> connectorDetails =
+        addReference(object, accountIdentifier, orgIdentifier, projectIdentifier, contextMap);
+    final Set<VisitedSecretReference> secretReferences = new HashSet<>();
```
Comment (acasian-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java:52]: nit `new HashSet<>(connectorDetails.size())`

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    return secretReferences;
+  }
+
+  private Set<IdentifierRef> getSecretRefsFromConnector(
+      String accountIdentifier, String orgIdentifier, String projectIdentifier, ConnectorInfoDTO connectorInfoDTO) {
+    if (connectorInfoDTO == null) {
+      return Collections.emptySet();
+    }
+
+    final Set<IdentifierRef> secretRefs = new HashSet<>();
```
Comment (acasian-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java:91]: nit: same as above

**Review by acasian-harness — APPROVED** [review]

**Comment by abosii-harness** [comment]
trigger review, gitleaks

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java (hunk 1 above)
Comment (abosii-harness) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/SecretConnectorRefExtractorHelper.java:52]: Will make these changes as part of the next PR

**Comment by abosii-harness** [comment]
trigger review

**Comment by abosii-harness** [comment]
trigger gitleaks

**Pull request merged**